substrate/client/authority-discovery/src/worker.rs

```diff
@@ -226,27 +238,32 @@ where
 			None => None,
 		};
 
-		// Make sure we have only valid peer ids in public addresses.
-		let public_addresses_only = {
-			let local_peer_id: Multihash = network.local_peer_id().into();
-
-			config.public_addresses_only.map(|addresses| {
-				addresses
-					.into_iter()
-					.map(|mut address| {
-						if let Some(multiaddr::Protocol::P2p(peer_id)) = address.iter().last() {
-							if peer_id != local_peer_id {
-								error!(
-									target: LOG_TARGET,
-									"Discarding invalid local peer ID in public address {address}.",
-								);
-								address.pop();
+		let published_addresses = match config.public_addresses_only {
+			Some(addresses) => {
+				PublishedAddresses::List({
+					// Make sure we have only valid peer ids in public addresses.
+					let local_peer_id: Multihash = network.local_peer_id().into();
+
+					addresses
+						.into_iter()
+						.map(|mut address| {
```

Comment:
small nit: If we `filter_map` these addresses and also add `multiaddr::Protocol::P2p(local_peer_id)` we could avoid a double allocation from the `.collect()` in `addresses_to_publish` for external addresses

Reply:
Not sure that I understand. Could you elaborate?

Reply:
We could group [ the section](https://github.com/paritytech/polkadot-sdk/pull/3657/files#diff-6ce157fc193b9691b2c38c118ee81244da354c65166edad5fb4b93beaa7b0f85R376-R385) that adds `/p2p/peer_id` with the [iteration](https://github.com/paritytech/polkadot-sdk/pull/3657/files#diff-6ce157fc193b9691b2c38c118ee81244da354c65166edad5fb4b93beaa7b0f85R372) through external addresses and do a `collect()` only once. 

And for the `PublishedAddresses::List` return them directly in `addresses_to_publish` if we make sure to add `/p2p/peer_id` sufix here

Reply:
Makes sense, done.